iceoryx_utils/include/iceoryx_utils/internal/relocatable_pointer/base_relocatable_ptr.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_UTILS_RELOCATABLE_POINTER_BASE_RELOCATABLE_PTR_HPP
+#define IOX_UTILS_RELOCATABLE_POINTER_BASE_RELOCATABLE_PTR_HPP
+
+#include <cstdint>
+#include <limits>
+
+namespace iox
+{
+namespace rp
+{
+/// @brief pointer class to use when pointer and pointee are located in the same shared memory segment
+/// We can have the following scenario:
+/// Pointer p points to object X of type T and both are stored in shared memory segment S.
+///
+/// Shared Memory   S:    p                  X
+///                       |__________________^
+/// App1            a1    b1                 c1
+/// App2            a2    b2                 c2
+///
+/// Let a1, b1, c1 be the addresses of segment S, pointer p and object X in application 1 and similarly a2, b2, and c2
+/// in application 2. If application 2 maps the memory differently they will be shifted by some common offset d
+/// depending on the individual memory mapping: a2=a1+d, b2=b1+d, c2=c1+d
+/// This is why storing a raw pointer to X will not be sufficient, the value of c1 will not point to X in application 2.
+/// However, storing the difference between the location of p and X will work since it is an invariant in both address
+/// spaces.
+class BaseRelocatablePointer
+{
+    template <typename T>
+    friend class RelocatablePointer;
+
+  public:
+    using offset_t = std::ptrdiff_t;
+
+    /// @brief default constructs a logical nullptr
+    BaseRelocatablePointer() noexcept;
+
+    /// @brief creates a relocatable pointer pointing to ptr
```

Comment:
same pointee as ptr (pointer and relocatable pointer are different entities but they are equivalent wrt. what they point to ...)